File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0;
        int maxLength = 0;
        Set<Character> charSet = new HashSet<>();
        for (int right=0; right<s.length(); right++) {
            if (charSet.contains(s.charAt(right))) {
                while(s.charAt(left)!=s.charAt(right)) {
                    charSet.remove(s.charAt(left));
                    left++;
                }
                charSet.remove(s.charAt(left));
                left++;
                charSet.add(s.charAt(right));
            } else {
                charSet.add(s.charAt(right));
                maxLength = Math.max(maxLength, right-left+1);
            }
        }
        return maxLength;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_set = set()
        max_length = 0
        left = 0
        for right in range(len(s)) :
            if s[right] in char_set :
                while s[left] != s[right] :
                    char_set.remove(s[left])
                    left+=1
                char_set.remove(s[left])
                left+=1
                char_set.add(s[right])
            else :
                char_set.add(s[right])
                max_length = max(max_length,right-left+1)
        return max_length
        